clang/tools/driver/CacheLauncherMode.cpp

```diff
@@ -30,15 +32,50 @@ static bool isSameProgram(StringRef clangCachePath, StringRef compilerPath) {
   return CacheStat.getUniqueID() == CompilerStat.getUniqueID();
 }
 
+static bool shouldCacheInvocation(ArrayRef<const char *> Args,
+                                  IntrusiveRefCntPtr<DiagnosticsEngine> Diags) {
+  SmallVector<const char *, 128> CheckArgs(Args.begin(), Args.end());
+  // Make sure "-###" is not present otherwise we won't get an object back.
+  CheckArgs.erase(
+      llvm::remove_if(CheckArgs, [](StringRef Arg) { return Arg == "-###"; }),
+      CheckArgs.end());
+  CreateInvocationOptions Opts;
+  Opts.Diags = Diags;
+  std::shared_ptr<CompilerInvocation> CInvok =
+      createInvocation(CheckArgs, std::move(Opts));
+  if (!CInvok)
+    return false;
+  if (CInvok->getLangOpts()->Modules) {
+    Diags->Report(diag::warn_clang_cache_disabled_caching)
+        << "-fmodules mode is enabled";
+    return false;
+  }
+  if (CInvok->getLangOpts()->AsmPreprocessor) {
+    Diags->Report(diag::warn_clang_cache_disabled_caching)
+        << "assembler language mode is enabled";
+    return false;
+  }
+  return true;
+}
+
 Optional<int>
 clang::handleClangCacheInvocation(SmallVectorImpl<const char *> &Args,
                                   llvm::StringSaver &Saver) {
   assert(Args.size() >= 1);
 
   auto DiagsConsumer = std::make_unique<TextDiagnosticPrinter>(
       llvm::errs(), new DiagnosticOptions(), false);
-  DiagnosticsEngine Diags(new DiagnosticIDs(), new DiagnosticOptions());
+  IntrusiveRefCntPtr<DiagnosticsEngine> DiagsPtr(
+      new DiagnosticsEngine(new DiagnosticIDs(), new DiagnosticOptions()));
+  DiagnosticsEngine &Diags = *DiagsPtr;
   Diags.setClient(DiagsConsumer.get(), /*ShouldOwnClient=*/false);
+  if (::getenv("LLVM_CACHE_ERROR_IF_NON_CACHING")) {
+    Diags.setDiagnosticGroupWarningAsError("compile-job-cache", true);
+  }
+  if (::getenv("LLVM_CACHE_HIDE_NON_CACHING_WARNING")) {
+    Diags.setSeverityForGroup(diag::Flavor::WarningOrError, "compile-job-cache",
```

Comment:
Same as this one, need to update DiagGroup

Reply:
Updated.